src/operator/contrib/two_bit_quantize-inl.h

```diff
@@ -0,0 +1,340 @@
+}
+
+inline bool Create2BitArrayShape(const nnvm::NodeAttrs& attrs,
+                                 std::vector<TShape> *in_attrs,
+                                 std::vector<TShape> *out_attrs) {
+  // 0. input array
+  CHECK_EQ(in_attrs->size(), 1U);
+  // 0. output array
+  CHECK_EQ(out_attrs->size(), 1U);
+  // check input
+  CHECK(!shape_is_none(in_attrs->at(0)));
+  // output
+  int shape = in_attrs->at(0).Size() % 16 == 0 ?
+                    in_attrs->at(0).Size() / 16 + 3:
+                    in_attrs->at(0).Size() / 16 + 4;
+  SHAPE_ASSIGN_CHECK(*out_attrs, 0, TShape{shape});
+  return true;
+}
+
+inline bool Create2BitArrayType(const nnvm::NodeAttrs &attrs,
+                                std::vector<int> *in_attrs,
+                                std::vector<int> *out_attrs) {
+  // 0. input array
+  CHECK_EQ(in_attrs->size(), 1U);
+  // 0. output array
+  CHECK_EQ(out_attrs->size(), 1U);
+  // check input
+  CHECK_EQ((*in_attrs)[0], mshadow::kFloat32)
+    << "`create_2bit_` only supports float32 input for now";
+  TYPE_ASSIGN_CHECK(*out_attrs, 0, mshadow::kFloat32);
+  return true;
+}
+
+struct init_threshold_2bit {
+  MSHADOW_XINLINE static void Map(int i,
+                                  float *out,
+                                  const float neg_threshold,
+                                  const float pos_threshold,
+                                  int size) {
+    // The first two elements in output are thresholds
+    // The third element is the original size of the array
+    out[0] = neg_threshold;
+    out[1] = pos_threshold;
+    // TODO(huilgolr) check potential problem here?
+    out[2] = static_cast<float>(size);
+  }
+};
+
+struct quantize_2bit {
+  MSHADOW_XINLINE static void Map(int block_id,
+                                  int grad_size,
+                                  float *out,
+                                  float *grad,
+                                  float *residual,
+                                  const float neg_threshold,
+                                  const float pos_threshold) {
+    float* compr_block = out + block_id;
+    // init to 0
+    *compr_block = 0;
+    // start and end are indices in original grad array
+    int start = block_id*16;
+    int end = (start+16 <= grad_size) ? start+16 : grad_size;
+    char* block_ptr = reinterpret_cast < char* > (compr_block);
+    for (int i=start; i < end; i++){
+      char* curr_byte = block_ptr + (i-start)/4;
+      float curr_value = grad[i] + residual[i];
+      if (curr_value >= pos_threshold) {
+        residual[i] = curr_value - pos_threshold;
+        // set data to 10
+        (*curr_byte) |= (2u << (6-((i%4)*2)));
+      } else if (curr_value <= neg_threshold) {
+        residual[i] = curr_value - neg_threshold;
+        // set data to 01
+        (*curr_byte) |= (1u << (6-((i%4)*2)));
+      } else {
+        // leave data as 00
+        residual[i] = curr_value;
```

Comment:
identified a bug here, need to add to existing threshold. 
will fix and add a test case for this.